File: ios/Khrysalis/com/lightningkite/khrysalis/observables/binding/RatingBar.binding.actual.swift
//Stub file made with Khrysalis 2 (by Lightning Kite)
import Foundation


//--- RatingBar.bind(Int, MutableObservableProperty<Int>)
public extension UIRatingBar {
    func bind(_ stars: Int, _ observable: MutableObservableProperty<Int>) -> Void {
        self.settings.totalStars = Int(stars)
        self.settings.fillMode = .full

        var suppress = false
        observable.subscribeBy { (value) in
            guard !suppress else { return }
            suppress = true
            self.rating = Double(value)
            suppress = false
        }.until(self.removed)
        self.didTouchCosmos = { rating in
            guard !suppress else { return }
            suppress = true
            observable.value = Int(rating)
            suppress = false
        }
    }
    func bind(stars: Int, observable: MutableObservableProperty<Int>) -> Void {
        return bind(stars, observable)
    }
}

//--- RatingBar.bind(Int, ObservableProperty<Int>)
public extension UIRatingBar {
    func bind(_ stars: Int, _ observable: ObservableProperty<Int>) -> Void {
        self.settings.totalStars = Int(stars)
        self.settings.fillMode = .full

        observable.subscribeBy { (value) in
            self.rating = Double(value)
        }.until(self.removed)
    }
    func bind(stars: Int, observable: ObservableProperty<Int>) -> Void {
        return bind(stars, observable)
    }
}

//--- RatingBar.bindFloat(Int, MutableObservableProperty<Float>)
public extension UIRatingBar {
    func bindFloat(_ stars: Int, _ observable: MutableObservableProperty<Float>) -> Void {
        self.settings.totalStars = Int(stars)
        self.settings.fillMode = .precise

        var suppress = false
        observable.subscribeBy { (value) in
            guard !suppress else { return }
            suppress = true
            self.rating = Double(value)
            suppress = false
        }.until(self.removed)
        self.didTouchCosmos = { rating in
            guard !suppress else { return }
            suppress = true
            observable.value = Float(rating)
            suppress = false
        }
    }
    func bindFloat(stars: Int, observable: MutableObservableProperty<Float>) -> Void {
        return bindFloat(stars, observable)
    }
}

//--- RatingBar.bindFloat(Int, ObservableProperty<Float>)
public extension UIRatingBar {
    func bindFloat(_ stars: Int, _ observable: ObservableProperty<Float>) -> Void {
        self.settings.totalStars = Int(stars)
        self.settings.fillMode = .precise

        observable.subscribeBy { (value) in
            self.rating = Double(value)
        }.until(self.removed)
    }
    func bindFloat(stars: Int, observable: ObservableProperty<Float>) -> Void {
        return bindFloat(stars, observable)
    }
}

File: android/src/main/java/com/lightningkite/khrysalis/observables/binding/RatingBar.binding.actual.kt
package com.lightningkite.khrysalis.observables.binding

import android.widget.RatingBar
import com.lightningkite.khrysalis.observables.*
import com.lightningkite.khrysalis.rx.removed
import com.lightningkite.khrysalis.rx.until

/**
 *
 * Binds the rating bar to the observable provided, as well allows you to provide how
 * many stars the rating is out of. Any changes to the rating bar will update the observable value,
 * as well any change to the observable directly will update the rating bar.
 *
 * Example
 * val rating:MutableObservableProperty<Int> = StandardObservableProperty(5)
 * ratingBar.bind(5, rating)
 *
 */

fun RatingBar.bind(
    stars: Int,
    observable: MutableObservableProperty<Int>
) {
    this.max = stars
    this.numStars = stars
    this.incrementProgressBy(1)

    var suppress = false
    observable.subscribeBy { value ->
        if (!suppress) {
            suppress = true
            this.progress = value
            suppress = false
        }
    }.until(this.removed)
    this.onRatingBarChangeListener = object : RatingBar.OnRatingBarChangeListener {
        override fun onRatingChanged(p0: RatingBar, p1: Float, p2: Boolean) {
            if (!suppress) {
                suppress = true
                observable.value = p0.progress
                suppress = false
            }
        }
    }

}

/**
 *
 * Binds the rating bar to the observable provided, as well allows you to provide how
 * many stars the rating is out of. The value cannot be changed from the rating bar itself,
 * though any change to the observable directly will manifest in the rating bar.
 *
 * Example
 * val rating: ObservableProperty<Int> = StandardObservableProperty(5)
 * ratingBar.bind(5, rating)
 *
 */

fun RatingBar.bind(
    stars: Int,
    observable: ObservableProperty<Int>
) {
    this.max = stars
    this.numStars = stars
    this.setIsIndicator(true)

    observable.subscribeBy { value ->
        this.progress = value
    }.until(this.removed)
}



/**
 *
 * Binds the rating bar to the observable provided, as well allows you to provide how
 * many stars the rating is out of. Any changes to the rating bar will update the observable value,
 * as well any change to the observable directly will update the rating bar..
 *
 * Example
 * val rating: MutableObservableProperty<Float> = StandardObservableProperty(5.0f)
 * ratingBar.bind(5.0f, rating)
 *
 */

fun RatingBar.bindFloat(
    stars: Int,
    observable: MutableObservableProperty<Float>
) {
    this.numStars = stars
    this.stepSize = 0.01f

    var suppress = false
    observable.subscribeBy { value ->
        if (!suppress) {
            suppress = true
            this.rating = value
            suppress = false
        }
    }.until(this.removed)
    this.onRatingBarChangeListener = object : RatingBar.OnRatingBarChangeListener {
        override fun onRatingChanged(p0: RatingBar, p1: Float, p2: Boolean) {
            if (!suppress) {
                suppress = true
                observable.value = p1
                suppress = false
            }
        }
    }

}


/**
 *
 * Binds the rating bar to the observable provided, as well allows you to provide how
 * many stars the rating is out of. The value cannot be changed from the rating bar itself,
 * though any change to the observable directly will manifest in the rating bar.
 *
 * Example
 * val rating: ObservableProperty<Float> = StandardObservableProperty(5.0f)
 * ratingBar.bind(5.0f, rating)
 *
 */

fun RatingBar.bindFloat(
    stars: Int,
    observable: ObservableProperty<Float>
) {
    this.numStars = stars
    this.setIsIndicator(true)

    observable.subscribeBy { value ->
        this.rating = value
    }.until(this.removed)
}
